add api documentation for crowbar-init's upgrade related api

MaximilianMeister opened the pull request:
as crowbar-init doesn't support apipie there is now now way to nicely
document the endpoints that we have there.

and to not have the documentation in yet another place it might be
better to include it directly in the crowbar api docs.

the downside of that is that the methods don't live in the crowbar
codebase, and therefore need to be faked by adding an empty dummy
method

rsalevsky commented on crowbar_framework/app/controllers/api/upgrade_controller.rb:
```diff
@@ -170,4 +170,67 @@ def cancel
+  protected
+
+  api :POST, "/api/upgrade/new",
+    "Initialization of Crowbar during upgrade with creation of a new database"
```
Can you add a hint that crowbar-init needs be running for this command?

santiph commented on crowbar_framework/app/controllers/api/upgrade_controller.rb:
```diff
@@ -170,4 +170,72 @@ def cancel
+    NOTE: It is only possible to use this endpoint during the stage where crowbar-init is running."
+  header "Accept", "application/vnd.crowbar.v2.0+json", required: true
+  param :username, String, desc: "Username"
+  param :password, String, desc: "Password"
```
For usarname and password:
Min length? Max length? Special characters required/allowed? Please update

santiph commented on crowbar_framework/app/controllers/api/upgrade_controller.rb:
```diff
@@ -170,4 +170,72 @@ def cancel
+  protected
+
+  api :POST, "/api/upgrade/new",
+    "Initialization of Crowbar during upgrade with creation of a new database.
```
It sounds like the database creation is part of this endpoint.
If that's not the case, please update the description

MaximilianMeister replied:
it is the case (at least for now)

santiph commented on crowbar_framework/app/controllers/api/upgrade_controller.rb:
```diff
@@ -170,4 +170,72 @@ def cancel
+  param :username, String, desc: "External database username"
+  param :password, String, desc: "External database password"
+  param :database, String, desc: "Database name"
+  param :host, String, desc: "External database host"
```
Ipv4? Ipv6? Hostname? Domain? What is supported here?

santiph commented on crowbar_framework/app/controllers/api/upgrade_controller.rb:
```diff
@@ -170,4 +170,72 @@ def cancel
+  header "Accept", "application/vnd.crowbar.v2.0+json", required: true
+  param :username, String, desc: "External database username"
+  param :password, String, desc: "External database password"
+  param :database, String, desc: "Database name"
```
For Username, Password and database name:
Min length? Max length? Special characters required/allowed? Please update

santiph commented on crowbar_framework/app/controllers/api/upgrade_controller.rb:
```diff
@@ -170,4 +170,72 @@ def cancel
+    }
+  }
+  '
+  error 406, "Connection to external database failed"
```
According to https://etherpad.nue.suse.com/p/cloud-upgrade-error-handling
there are several cases that could produce a failure in the connection:

- Host not found
- Incorrect port
- Wrong credentials: Return "wrong credentials", not "wrong user name" or "wrong password"
- Wrong database name

MaximilianMeister replied:
I'll add that to the error description

santiph commented on crowbar_framework/app/controllers/api/upgrade_controller.rb:
```diff
@@ -170,4 +170,72 @@ def cancel
+  param :password, String, desc: "Password"
+  api_version "2.0"
+  example '
+  {
```
Didn't we agreed on split the endpoints between Database creation/connection and Crowbar initialisation?

MaximilianMeister replied:
yes, we have to do this in a follow up, as this here is just to document what we have now.

MaximilianMeister commented:
@santiph I addressed your comments, and added a regex (from https://github.com/crowbar/crowbar-client/blob/master/lib/crowbar/client/mixin/database.rb) instead of a simple datatype to the database params

![crowbar-init-doc-regex](https://cloud.githubusercontent.com/assets/5364817/19261913/6484d076-8f94-11e6-826a-0eeedda22c17.png)

santiph commented:
@MaximilianMeister Its good to have the regex validations, but we still need to get documentation for the parameters limitations and know the minimum and max length, if there is any special character required (number, symbols, upper or lowercase letters) so we can show a better message to the user than "Your input doesn't match ^(([a-zA-Z0-9]|[a-zA-Z0-9][a-zA-Z0-9\-]*[a-zA-Z0-9])\.)*([A-Za-z0-9]|[A-Za-z0-9][A-Za-z0-9\-]*[A-Za-z0-9])$", which is a terrible UX

MaximilianMeister commented:
> Its good to have the regex validations, but we still need to get documentation for the parameters limitations and know the minimum and max length, if there is any special character required (number, symbols, upper or lowercase letters) so we can show a better message to the user than "Your input doesn't match ^(([a-zA-Z0-9]|[a-zA-Z0-9][a-zA-Z0-9-][a-zA-Z0-9]).)([A-Za-z0-9]|[A-Za-z0-9][A-Za-z0-9-]*[A-Za-z0-9])$", which is a terrible UX

not sure how you want to do that, even if I add the text here in the apidoc, you will likely not read or parse it from the UI. don't you have to include that human readable form in the UI code?

santiph commented:
It's not to parse it, but to know it instead. We need to have those validation rules stored in the UI, but it's important for us to do so that we have documentation in place indicating such restrictions

MaximilianMeister commented:
@santiph right, it wouldn't hurt and looks more user friendly. thanks, fixed

![param_desc](https://cloud.githubusercontent.com/assets/5364817/19300298/d55570ce-9059-11e6-95f3-f078f0326e3d.png)

santiph commented on crowbar_framework/app/controllers/api/upgrade_controller.rb:
```diff
@@ -170,4 +170,115 @@ def cancel
+    desc: "Password
+      Min length: 4
+      Max length: 63
+      Alphanumeric and special characters
```
What are considered special characters here?
ä, ö, ü, ë, ñ?
Symbols?
Is there any restriction or everything is supported?

santiph commented on crowbar_framework/app/controllers/api/upgrade_controller.rb:
```diff
@@ -170,4 +170,115 @@ def cancel
+    desc: "External database username
+      Min length: 4
+      Max length: 63
+      Only alphanumeric characters or underscores
```
and/or*

santiph commented on crowbar_framework/app/controllers/api/upgrade_controller.rb:
```diff
@@ -170,4 +170,115 @@ def cancel
+      Min length: 4
+      Max length: 63
+      Numbers and period characters (only IPv4)
+      Hostnames:
```
Hostname vs Domains? Is there any difference we should care about?

santiph commented:
Thanks @MaximilianMeister 
API Documentation looks great now :)
I left a couple of comments more. I think it will be all before approve and merge it

Itxaka commented on crowbar_framework/app/controllers/api/upgrade_controller.rb:
```diff
@@ -170,4 +170,115 @@ def cancel
+       alphanumeric characters and hyphens
+       cannot start/end with digits or hyphen",
+    required: true
+  param :port, /(?=^.{1,5}$)(?=^[0-9]*$)/,
```
Dont we need to validate that this is a number lower than 65535?

MaximilianMeister replied:
good point, though it would have been good to know that during https://github.com/crowbar/crowbar-client/pull/105 ;)

but here we are documenting the actual state, so first would be the fix in crowbar-client, and then another follow up for the api docs here.

santiph replied:
+1

MaximilianMeister replied:
see https://github.com/crowbar/crowbar-client/pull/113

MaximilianMeister commented:
![](https://github.trello.services/images/mini-trello-icon.png) [(5) S49P9 #upgrade Synchronize API responses with the Angular app](https://trello.com/c/3TALQwDA/121-5-s49p9-upgrade-synchronize-api-responses-with-the-angular-app)
